Crossover2 and weightmapping

AndrewLouw opened the pull request:
All finifhed everything works as intended

AndrewLouw committed:
crossover2 function added

Crossover 2 takes a random city and finds it's position in the parent's
order vector, then choses one parent randomly but weighted based on
their performance and places thic city in the output vector in this
position. If the first choice parent's position is taken alread the
second choice is used, if both are taken it is placed randomly at the
end. Mutations prevent the whole population becoming identical and
introduce new samples occasionall. Works well with weightmapping -
replaces crossover in sketch.

AndrewLouw committed:
Weightmapping

the whole population is displayed at once in the upper figure. The
thickness of the lines are proportional to how many networks take that
route. Making it easy to see if the population is evolving or randomly
guessing.

AndrewLouw committed:
Tweaks

comparison now squares to give better one a bigger advantage, mutation
is now recursive to increase double mutation chance.

AndrewLouw committed:
Tweaks

mutate is a random swap not shuffle, testing more cities and different
probabilities.

AndrewLouw committed:
Mutation function moved to Sketch

Mutate function wasn't running in DNA but works in skech, more features
have been added and some values fiddled.

AndrewLouw committed:
Fixed mutate added apocolypse

Mutate was being called but not returning a value, ultimately leading to
crasthes but more annoyingly not actually modifying the order. also now
every1000 generations there are 5 generataions of extream mutation to
help keep things progressing. The best ever is always in the population
now also

AndrewLouw committed:
Valuechange and bug fix

used the wrong variable name in one place

shiffman commented:
Wow, these are wonderful improvements! I like to keep my examples simple -- this one, for example, is meant to just demonstrate a GA without any crossover at all. I see two options:

1. Create a new example (separate from my first two) with this code in this repo (03_TSP. . )
2. Create a new example in your own repo with this code and I'll link outwards from my README.

Do you have a preference?

AndrewLouw commented:
I don't have a preference, but your Github gets much more traffic and I'm only just learning how to use mine, so I think it's best if you make a new example.

AndrewLouw commented:
I don't know if you ever do followup videos, or if I've anything new enough to make it worth it but feel free to use this code.

AndrewLouw committed:
Parenting and value changes

Added 10000th roor parenting this very mild bias towards the better
parent means slightly inefficent networks can form a breeding population
allowing improvements which require muliple mutations to be found -
anything is better than equal parenting however.
apocalypse spelling correction... and happens more often with more
mutations but is shorter.
Maximum stack exceeded error prevented (was low probability)
Now 'solves' for 50 cities in about 5 minutes

AndrewLouw committed:
Values

AndrewLouw commented:
These last couple look at the parenting selection and vastly improve performance by increasing the propagation of mildly inefficient DNA
